File: javascript/resistor-color-trio/resistor-color-trio.js
//
// This is only a SKELETON file for the 'Resistor Color Trio' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class ResistorColorTrio {
  constructor(colors) {
    this.colors = colors;
    this.colorValues = {
      'black': 0,
      'brown': 1,
      'red': 2,
      'orange': 3,
      'yellow': 4,
      'green': 5,
      'blue': 6,
      'violet': 7,
      'grey': 8,
      'white': 9
    };
    
    // Validate colors
    for (const color of colors) {
      if (!(color in this.colorValues)) {
        throw new Error('invalid color');
      }
    }
  }

  get label() {
    // Get the first two digits
    const firstDigit = this.colorValues[this.colors[0]];
    const secondDigit = this.colorValues[this.colors[1]];
    
    // Calculate the main value (first two digits)
    const mainValue = firstDigit * 10 + secondDigit;
    
    // Get the number of zeros to add (from the third color)
    const zeros = this.colorValues[this.colors[2]];
    
    // Calculate the total value
    let value = mainValue * Math.pow(10, zeros);
    let unit = 'ohms';
    
    // Convert to kiloohms if value is 1000 or more
    if (value >= 1000) {
      value = value / 1000;
      unit = 'kiloohms';
    }
    
    // Return the formatted label
    return `Resistor value: ${value} ${unit}`;
  }
}

File: java/resistor-color-trio/src/main/java/ResistorColorTrio.java
class ResistorColorTrio {
    String label(String[] colors) {
        // Map colors to their numeric values
        int firstDigit = getColorValue(colors[0]);
        int secondDigit = getColorValue(colors[1]);
        int zeros = getColorValue(colors[2]);
        
        // Calculate the main value
        int mainValue = firstDigit * 10 + secondDigit;
        
        // Calculate the total value
        long totalValue = mainValue * (long)Math.pow(10, zeros);
        
        // Format the output with the appropriate metric prefix
        return formatOutput(totalValue);
    }
    
    private int getColorValue(String color) {
        switch (color.toLowerCase()) {
            case "black": return 0;
            case "brown": return 1;
            case "red": return 2;
            case "orange": return 3;
            case "yellow": return 4;
            case "green": return 5;
            case "blue": return 6;
            case "violet": return 7;
            case "grey": return 8;
            case "white": return 9;
            default: throw new IllegalArgumentException("Invalid color: " + color);
        }
    }
    
    private String formatOutput(long value) {
        if (value < 1000) {
            return value + " ohms";
        } else if (value < 1_000_000) {
            return (value / 1000) + " kiloohms";
        } else if (value < 1_000_000_000) {
            return (value / 1_000_000) + " megaohms";
        } else {
            return (value / 1_000_000_000) + " gigaohms";
        }
    }
}
